src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -31,10 +32,25 @@ internal class SNIProxy
         /// </summary>
         /// <param name="sspiClientContextStatus">SSPI client context status</param>
         /// <param name="receivedBuff">Receive buffer</param>
-        /// <param name="sendBuff">Send buffer</param>
         /// <param name="serverName">Service Principal Name buffer</param>
-        /// <returns>SNI error code</returns>
-        internal static void GenSspiClientContext(SspiClientContextStatus sspiClientContextStatus, byte[] receivedBuff, ref byte[] sendBuff, byte[][] serverName)
+        /// <returns>Memory for response</returns>
+        internal static IMemoryOwner<byte> GenSspiClientContext(SspiClientContextStatus sspiClientContextStatus, ReadOnlyMemory<byte> receivedBuff, byte[][] serverName)
+        {
+            // TODO: this should use ReadOnlyMemory all the way through
+            var array = ArrayPool<byte>.Shared.Rent(receivedBuff.Length);
```

Comment:
Please consider when switching to `ReadOnlyMemory` the problem that the array returned by `ArrayPool<byte>.Shared.Rent` might me non-null-initialized and larger than the length specified. That causes sometimes problems during runtime like mentioned in [in this issue](https://github.com/dotnet/SqlClient/issues/2441).

Reply:
I think this implementation will actually fix that issue as is, as it returns a Memory<byte> rather than the actual array so the extra entries wouldn't be accessible.

Reply:
But aren't we still passing an instance of `byte[]` down to the `GSSAPI.InitSecContext` which might have some wrong bytes from the `ArrayPool`?
Cause the new `MemoryStream` is created from the `outSecurityBuffer`, but the bug is caused by the `receivedBuff`, which after `receivedBuff.CopyTo(array);` might have some non-zero-bytes after the copied bytes.

Reply:
I see - I've update the array to be cleared out when renting